railties/lib/rails/test_unit/runner.rb

```diff
@@ -105,8 +105,12 @@ def list_tests(patterns)
 
           def escape_declarative_test_filter(filter)
             if filter.is_a?(String) && !filter.start_with?("test_")
-              filter = "test_#{filter}" unless regexp_filter?(filter)
-              filter = filter.gsub(/\s+/, "_")
+              new_filter = filter.gsub(/ (?!_)/, "[ _]")
```

Comment:
`escape_declarative_test_filter` gets called many times, I think it gets called once for every test class that's been loaded or something like that. Without the negative lookahead, the regex expands enormously:

```
/foo bar/ => /foo[ _]bar/
 /foo[ _]bar/ =>  /foo[[ _]_]bar/
/foo[[ _]_]bar/ => /foo[[[ _]_]_]bar/
```

The negative lookahead prevents that.

Reply:
You can check if this is happening by adding `puts "filter becomes #{filter}"` before the return at the end of the method.

Reply:
Interesting!  I think that happens because Minitest [calls `run` twice](https://github.com/minitest/minitest/blob/a9fa045044b4210cfd21a512b06d1a4527d709ba/lib/minitest.rb#L182-L183) and we [apply `compose_filter` in `run`](https://github.com/rails/rails/blob/f95c0b7e96eb36bc3efc0c5beffbb9e84ea664e4/railties/lib/rails/test_unit/line_filtering.rb#L8).  Using `gsub(/\s+/, ...)` should avoid that problem, but it's important to be aware of, so I have added a comment to `escape_declarative_test_filter`.

Reply:
It was getting called *lots* of times, many more than 2. Anyway, your solution looks good.